### DISCUSSION TITLE
How to typecast field in where parameters

### DISCUSSION THREAD

**Discussion by sebapatr96** [mikro-orm/mikro-orm#4774] [Q&A] [answered]
Hello,

is it possible in v6 to typecast property in where parameters? I have an Entity `Ip` with column `address` of type `inet` in postgres, and I want to do filtering based on the address column but in text type, not inet

I would like to do equivalent of this sql query in EM api

```sql
 select "i0".* from "public"."ip" as "i0" where cast(i0.address as text) = '192.168.0.0'
```

(I would like to also use ilike etc, not only equal which could get fixed just by adding /32 notation at the end of ip string btw)

I have tried this solution

```typescript
const ips = await em.find(Ip, {
  ['cast(i0.address as text)']: '192.168.0.0',
} as any);
```

But it does not work because it wrongly converts the cast operator in sql

```sql
select "i0".* from "public"."ip" as "i0" where "cast(i0"."address" as "text)" = '192.168.0.0' - syntax error at or near "as"
```

Is there please any solution to this? I also tried `::text` but with the same issue
```
where "i0"."address::text" = '192.168.0.0' - column i0.address::text does not exist
```

**Comment by B4nan** [comment, marked as answer]
Weird, we have pretty much the same example in the tests and it works fine:

https://github.com/mikro-orm/mikro-orm/blob/master/tests/EntityManager.postgre.test.ts#L1575

(the `expr` helper will help you with getting around strict type checks, so you won't need the `any` cast, but it has no effect on runtime, I don't think it's about that)

In v6 you should be able to do this:

```ts
const ips = await em.find(Ip, {
  [sql`cast(i0.address as text)`]: '192.168.0.0',
});
```

**Reply by sebapatr96** [reply]
Thanks this works!

**Reply by sebapatr96** [reply]
it seems like its working differently based on context where its used 🤔 If I use it as a relation where argument like

```typescript
await em.find(User, {
  ips: {
    [raw(alias => `cast(${alias}.address as text)`)]: '192.168.0.0'
  }
}}
```

(I tried the raw helper function this time because it provides alias, but the error result is the same as if I would use sql )
```
[sql`cast(i1.address as text)`]: '192.168.0.0'
```

It creates this raw sql argument `"[raw]: cast(i1"."address" as "text) (#0)" = '192.168.0.0'` which throws error in postgres

**Reply by B4nan** [reply]
Can you provide a repro for that?

**Reply by sebapatr96** [reply]
I was trying to reproduce it in separated project with extracted entities but without luck, anyway it seems like in my main project it has something to do with eager loaded relation 🤔 when i removed eager it started working but im unable to reproduce it in extracted project. Idk will look into it tomorrow

**Reply by sebapatr96** [reply]
Ok I was able to reproduce it, this time it failed on orderBy, but the error is same. Here is the code

dependencies
```
"@mikro-orm/cli": "6.0.0-dev.114",
"@mikro-orm/core": "6.0.0-dev.114",
"@mikro-orm/migrations": "6.0.0-dev.114",
"@mikro-orm/postgresql": "6.0.0-dev.114",
```

Entities
```typescript
@Entity()
export class User extends BaseEntity {
  @PrimaryKey()
  readonly id: number;

  @OneToMany(() => Server, x => x.user)
  servers = new Collection<Server>(this);
}

@Entity()
export class Server extends BaseEntity {
  @PrimaryKey()
  readonly id: number;

  @ManyToOne(() => User, { ref: true })
  user: Ref<User>;

  @OneToMany(() => Ip, x => x.server)
  ips = new Collection<Ip>(this);
}

@Entity()
export class Ip extends BaseEntity {
  @PrimaryKey()
  readonly id: number;

  @Property({ columnType: 'inet' })
  address: string;

  @ManyToOne(() => Server, { ref: true })
  server: Ref<Server>;
}
```

Query
```typescript
await orm.em.fork().find(
  User,
  {
    servers: {
      ips: {
        [raw(alias => `cast(${alias}.address as text)`)]: {
          $ilike: '%192.168%',
        },
      },
    },
  },
  {
    populate: ['servers', 'servers.ips'],
    orderBy: {
      servers: {
        ips: {
          [raw(alias => `cast(${alias}.address as text)`)]: 'ASC',
        },
      },
    },
  },
)
```
